### ISSUE TITLE
[Installation Issues] - on arm-based Mac

### ISSUE THREAD

**Issue by ai-nikolai** [alfworld/alfworld#44]
## A bit of Context:
There were installation issues for the underlying Textworld for the arm based Mac, however, these are now resolved. (https://github.com/microsoft/TextWorld/issues/310) I.e. one can now install `Textworld` just using pip on mac. 

Installing on an arm based Mac is still an issue for Alfworld, however, as the first two steps of the Readme "reinstall" textworld the "old" way (before the issue was fixed on Textworld):
```bash
$ pip install https://github.com/MarcCote/downward/archive/faster_replan.zip
$ pip install https://github.com/MarcCote/TextWorld/archive/handcoded_expert_integration.zip
```

Concretely, the installation of Textworld (using the old way that happens automatically when installing these zip files) fails during installation at the Inform7 step:
```bash
      + tar xf I7_6M62_Linux_all.tar.gz
      + echo 'Installing Inform7 CLI'
      + cd inform7-6M62/
      + ./install-inform7.sh --prefix /private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-req-build-5u1gl9ar/textworld/thirdparty/inform7-6M62
      You should have the executable compilers for the arm64 architecture.
      Giving up!
```
As it did for Textworld, because there is no arm-compatible version of Inform7 (at the installation source: `http://emshort.com/inform-app-archive/6M62/I7-6M62-OSX-Interim.dmg`

----
## What could solve the problem:
1. @MarcCote please tell us how we can compile `faster_replan` & `handcoded_expert_integration` from scratch (as it is not clear where these archives come from - or where the source of these is).
2. Building new versions of `faster_replan` & `handcorded_expert_integration` that supports the arm based Mac (perhaps by releasing a new version that links to the latest textworld?)
3. Open to other ideas...

@MarcCote let us know how we can help, thanks a lot in advance.

**Comment by MarcCote** [comment]
@ai-nikolai I'm in the process of updating ALFWorld and its dependencies 
`fast-downward @ https://github.com/MarcCote/downward/archive/faster_replan.zip`
and
`textworld @ https://github.com/MarcCote/TextWorld/archive/handcoded_expert_integration.zip`

**Comment by ai-nikolai** [comment]
Thanks. Let me know when something should be tested.

**Comment by MarcCote** [comment]
@ai-nikolai. Can you try the following

    pip install alfworld==0.3.0rc2

**Comment by ai-nikolai** [comment]
Seems like it did not work out of the box (i.e. Creating a new environment and running `pip install alfworld==0.3.0rc2`. )
1. `Python 3.9.6`
2. `pip 23.3.1`

Installing textworld on its own (i.e. `pip install textworld` does work in the same environment.)

**The last few lines of the error:**
```bash
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed cleaning build dir for textworld
  Building wheel for fast-downward-textworld (setup.py) ... done
  Created wheel for fast-downward-textworld: filename=fast_downward_textworld-20.6.1-py3-none-any.whl size=438021 sha256=32e8bd310a11e0a32740e29209ab62ed43604b232e953e9344b10b2e96f49828
  Stored in directory: /Users/nikolai/Library/Caches/pip/wheels/e0/e9/6b/769674eb7ea3e34c41d486c483e70f5488cb017641a949bea9
Successfully built alfworld fast-downward-textworld
Failed to build textworld
ERROR: Could not build wheels for textworld, which is required to install pyproject.toml-based projects
```


**The full error:**
```bash
Building wheels for collected packages: alfworld, textworld, fast-downward-textworld
  Building wheel for alfworld (setup.py) ... done
  Created wheel for alfworld: filename=alfworld-0.3.0rc2-py3-none-any.whl size=352652 sha256=bd4d4b57cd5bee0d049230b2c5ea911d5caac71e0e6c309a4cb4e5bbc0018b49
  Stored in directory: /Users/nikolai/Library/Caches/pip/wheels/8d/44/87/ed6c2160a647a27734d6a288764811cc5f31105dc453db212e
  Building wheel for textworld (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [42 lines of output]
      /Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py:80: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
      !!
      
              ********************************************************************************
              Requirements should be satisfied by a PEP 517 installer.
              If you are using pip, you can try `pip install --use-pep517`.
              ********************************************************************************
      
      !!
        dist.fetch_build_eggs(dist.setup_requires)
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-install-qbe4zjm9/textworld_f3ac7994d1c44588a6c18dcd220b1513/setup.py", line 41, in <module>
          setup(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py", line 102, in setup
          _install_setup_requires(attrs)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py", line 75, in _install_setup_requires
          _fetch_build_eggs(dist)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py", line 80, in _fetch_build_eggs
          dist.fetch_build_eggs(dist.setup_requires)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/dist.py", line 662, in fetch_build_eggs
          return _fetch_build_eggs(self, requires)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/installer.py", line 38, in _fetch_build_eggs
          resolved_dists = pkg_resources.working_set.resolve(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 829, in resolve
          dist = self._resolve_dist(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 865, in _resolve_dist
          dist = best[req.key] = env.best_match(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1135, in best_match
          return self.obtain(req, installer)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1147, in obtain
          return installer(requirement)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/installer.py", line 106, in _fetch_build_egg_no_warn
          wheel.install_as_egg(dist_location)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/wheel.py", line 120, in install_as_egg
          self._install_as_egg(destination_eggdir, zf)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/wheel.py", line 128, in _install_as_egg
          self._convert_metadata(zf, destination_eggdir, dist_info, egg_info)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/wheel.py", line 173, in _convert_metadata
          os.rename(dist_info, egg_info)
      OSError: [Errno 66] Directory not empty: '/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-install-qbe4zjm9/textworld_f3ac7994d1c44588a6c18dcd220b1513/.eggs/cffi-1.16.0-py3.9-macosx-10.9-universal2.egg/cffi-1.16.0.dist-info' -> '/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-install-qbe4zjm9/textworld_f3ac7994d1c44588a6c18dcd220b1513/.eggs/cffi-1.16.0-py3.9-macosx-10.9-universal2.egg/EGG-INFO'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for textworld
  Running setup.py clean for textworld
  error: subprocess-exited-with-error
  
  × python setup.py clean did not run successfully.
  │ exit code: 1
  ╰─> [42 lines of output]
      /Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py:80: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
      !!
      
              ********************************************************************************
              Requirements should be satisfied by a PEP 517 installer.
              If you are using pip, you can try `pip install --use-pep517`.
              ********************************************************************************
      
      !!
        dist.fetch_build_eggs(dist.setup_requires)
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-install-qbe4zjm9/textworld_f3ac7994d1c44588a6c18dcd220b1513/setup.py", line 41, in <module>
          setup(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py", line 102, in setup
          _install_setup_requires(attrs)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py", line 75, in _install_setup_requires
          _fetch_build_eggs(dist)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/__init__.py", line 80, in _fetch_build_eggs
          dist.fetch_build_eggs(dist.setup_requires)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/dist.py", line 662, in fetch_build_eggs
          return _fetch_build_eggs(self, requires)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/installer.py", line 38, in _fetch_build_eggs
          resolved_dists = pkg_resources.working_set.resolve(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 829, in resolve
          dist = self._resolve_dist(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 865, in _resolve_dist
          dist = best[req.key] = env.best_match(
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1135, in best_match
          return self.obtain(req, installer)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1147, in obtain
          return installer(requirement)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/installer.py", line 106, in _fetch_build_egg_no_warn
          wheel.install_as_egg(dist_location)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/wheel.py", line 120, in install_as_egg
          self._install_as_egg(destination_eggdir, zf)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/wheel.py", line 128, in _install_as_egg
          self._convert_metadata(zf, destination_eggdir, dist_info, egg_info)
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/setuptools/wheel.py", line 173, in _convert_metadata
          os.rename(dist_info, egg_info)
      OSError: [Errno 66] Directory not empty: '/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-install-qbe4zjm9/textworld_f3ac7994d1c44588a6c18dcd220b1513/.eggs/cffi-1.16.0-py3.9-macosx-10.9-universal2.egg/cffi-1.16.0.dist-info' -> '/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-install-qbe4zjm9/textworld_f3ac7994d1c44588a6c18dcd220b1513/.eggs/cffi-1.16.0-py3.9-macosx-10.9-universal2.egg/EGG-INFO'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed cleaning build dir for textworld
  Building wheel for fast-downward-textworld (setup.py) ... done
  Created wheel for fast-downward-textworld: filename=fast_downward_textworld-20.6.1-py3-none-any.whl size=438021 sha256=32e8bd310a11e0a32740e29209ab62ed43604b232e953e9344b10b2e96f49828
  Stored in directory: /Users/nikolai/Library/Caches/pip/wheels/e0/e9/6b/769674eb7ea3e34c41d486c483e70f5488cb017641a949bea9
Successfully built alfworld fast-downward-textworld
Failed to build textworld
ERROR: Could not build wheels for textworld, which is required to install pyproject.toml-based projects
```

**Comment by ai-nikolai** [comment]
**Update: with pep517 it seems to work**
Following the above error. Creating a new environment, the following install did work (haven't tried running things yet):
```bash
pip install --use-pep517 alfworld==0.3.0rc2
```

**Comment by ai-nikolai** [comment]
Thanks for trying. Now this is throwing a new error:
```bash
Collecting textworld>=1.6.0rc2 (from textworld[pddl]>=1.6.0rc2->alfworld==0.3.0rc3)
  Downloading textworld-1.6.0rc2.tar.gz (710 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━ 710.5/710.5 kB 37.6 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [28 lines of output]
      Traceback (most recent call last):
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 366, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 311, in run_setup
          exec(code, locals())
        File "<string>", line 41, in <module>
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/__init__.py", line 103, in setup
          return distutils.core.setup(**attrs)
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 159, in setup
          dist.parse_config_files()
        File "/Users/nikolai/Desktop/Uni/1_software/reflexion/alfworld_runs/env_reflexion/lib/python3.9/site-packages/_virtualenv.py", line 23, in parse_config_files
          result = old_parse_config_files(self, *args, **kwargs)
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 627, in parse_config_files
          pyprojecttoml.apply_configuration(self, filename, ignore_option_errors)
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/config/pyprojecttoml.py", line 66, in apply_configuration
          config = read_configuration(filepath, True, ignore_option_errors, dist)
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/config/pyprojecttoml.py", line 127, in read_configuration
          validate(subset, filepath)
        File "/private/var/folders/y4/c4phbshn5739xq523fgf_cl40000gn/T/pip-build-env-m0pb63fs/overlay/lib/python3.9/site-packages/setuptools/config/pyprojecttoml.py", line 55, in validate
          raise ValueError(f"{error}\n{summary}") from None
      ValueError: invalid pyproject.toml config: `project`.
      configuration error: `project` must contain ['name'] properties
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

```

**Comment by MarcCote** [comment]
Ok. I fixed the pyproject.toml of TextWorld and made a new release candadite (1.6.0rc4) which should be picked up by

    pip install alfworld==0.3.0rc3

**Comment by MarcCote** [comment]
@ai-nikolai heads-up, I will be moving ahead with making releases of TextWorld and ALFWorld later today.

**Comment by ai-nikolai** [comment]
Thanks, I will make sure to test them. I will try to test them by end of the week. Thank you.

**Comment by MarcCote** [comment]
Should be addressed by #45.
    
    pip install alfworld>=0.3.1 

    alfworld-download
    alfworld-play-tw

Feel free to reopen the issue, otherwise.
